### PR TITLE
[mlir][tensor] Fix off-by-one error in ReshapeOpsUtils

### PR DESCRIPTION
**Pull request by vinayakdsci**
This patch fixes an off-by-one error in `mlir::getReassociationIndicesForCollapse()` that occurs when the last two dims of the source tensor satisfy the while loop.

This would cause an assertion failure due to out-of-bounds-access, which is now fixed.

### CODE REVIEW

**Comment by AmosLewis** [comment]
LGTM. Just tested with model.mlir in https://github.com/nod-ai/SHARK-ModelDev/issues/866. All good.

**Review by AmosLewis — APPROVED** [review]
LGTM. Just tested with model.mlir in https://github.com/nod-ai/SHARK-ModelDev/issues/866. All good.

File: mlir/test/Dialect/Tensor/canonicalize.mlir [hunk 1]
```diff
@@ -1251,6 +1251,29 @@ func.func @no_fold_expand_of_collapse_dynamic(%arg0 : tensor<?x?x?xf32>, %arg1:
...
+//       CHECK: #[[$MAP:.*]] = affine_map<()[s0] -> (s0 * 64)>
+// CHECK-LABEL: @compose_expand_of_collapse_last_two_dims
+//  CHECK-SAME: %[[ARG0:.+]]: tensor<?x64x1xf32>
+//       CHECK: %[[CONSTANT0:.+]] = arith.constant 0 : index
+//       CHECK: %[[CONSTANT384:.+]] = arith.constant 384 : index
+//       CHECK: %[[COLLAPSE:.+]] = tensor.collapse_shape %[[ARG0]] {{\[}}[0, 1, 2]] : tensor<?x64x1xf32> into tensor<?xf32>
+//       CHECK: %[[DIM:.+]] = tensor.dim %[[ARG0]], %[[CONSTANT0]] : tensor<?x64x1xf32>
+//       CHECK: %[[AFFAPPLY:.+]] = affine.apply #[[$MAP]]()[%[[DIM]]]
+//       CHECK: %[[DIVUI:.+]] = arith.divui %[[AFFAPPLY]], %[[CONSTANT384]] : index
+//       CHECK: %[[RESULT:.+]] = tensor.expand_shape %[[COLLAPSE]] {{\[}}[0, 1]] output_shape [%1, 384] : tensor<?xf32> into tensor<?x384xf32>
```
Comment (pashu123) [inline mlir/test/Dialect/Tensor/canonicalize.mlir:1272]: Please capture %1 here `output_shape [%1, 384]`.

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ mlir::getReassociationIndicesForCollapse(ArrayRef<int64_t> sourceShape,
       break;
 
     int64_t currTargetShape = targetShape[targetDim];
-    while (sourceDim < sourceShape.size() &&
+    while (sourceDim < sourceShape.size() - 1 &&
```
Comment (pashu123) [inline mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:50]:
```suggestion
    while (sourceDim < (sourceShape.size() - 1) &&
```

**Review by pashu123 — CHANGES_REQUESTED** [review]
We can wait for the original author to review it.

File: mlir/test/Dialect/Tensor/canonicalize.mlir (hunk 1 above)
Comment (vinayakdsci) [reply mlir/test/Dialect/Tensor/canonicalize.mlir]: @pashu123 thanks, addressed in the latest commit.

**Comment by vinayakdsci** [comment]
cc @pifon2a.

**Review by pifon2a — APPROVED** [review]
Thank you!

**Comment by vinayakdsci** [comment]
@pashu123 if the PR looks good, could you approve and merge?

**Review by pashu123 — APPROVED** [review]

**Pull request merged**